File: 송인재/프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.java
import java.util.*;

class Solution {
    public int solution(String[] friends, String[] gifts) {
        Map<String, Map<String, Integer>> giveAndTake = new HashMap<>();
        Map<String, Integer> giftDegree = new HashMap<>();
        Map<String, Integer> nextMonth = new HashMap<>();
        int max = 0;
        
        for (String friend : friends) {
            giveAndTake.put(friend, new HashMap<>());
            giftDegree.put(friend, 0);
            nextMonth.put(friend, 0);
        }
        
        for (String gift : gifts) {
            String[] giveTake = gift.split(" ");
            String give = giveTake[0];
            String take = giveTake[1];
            Map<String, Integer> friend = giveAndTake.get(give);
            
            friend.put(take, friend.getOrDefault(take, 0) + 1);
            giftDegree.put(give, giftDegree.get(give) + 1);
            giftDegree.put(take, giftDegree.get(take) - 1);
        }
        
        for (String giver : giveAndTake.keySet()) {
            for (String taker : giftDegree.keySet()) {
                if (taker.equals(giver)) {
                    continue;
                }
                
                int given = giveAndTake.get(giver)
                    .getOrDefault(taker, 0);
                int taken = giveAndTake.get(taker)
                    .getOrDefault(giver, 0);
                
                if (given > taken) {
                    nextMonth.put(giver, nextMonth.get(giver) + 1);
                }
                
                if (given == taken && giftDegree.get(giver) > giftDegree.get(taker)) {
                    nextMonth.put(giver, nextMonth.get(giver) + 1);
                }
            }

            max = Math.max(max, nextMonth.get(giver));
        }
        
        return max;
    }
}

File: 송인재/프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.kt
import java.util.*;

class Solution {
    fun solution(friends: Array<String>, gifts: Array<String>): Int {
        val giveAndTake = Array(friends.size) { IntArray(friends.size) {0} }
        val friendsMap = mutableMapOf<String, Int>()
        val giftDegree = IntArray(friends.size)
        val received = IntArray(friends.size)
        
        for ((i, friend) in friends.withIndex()) {
            friendsMap[friend] = i
        }
        
        for (record in gifts) {
            val content = record.split(" ")
            val gived = friendsMap[content[0]]!!
            val received = friendsMap[content[1]]!!
            giveAndTake[gived][received]++
            giftDegree[gived]++
            giftDegree[received]--
        }
        
        for ((i, row) in giveAndTake.withIndex()) {
            for ((j, given) in row.withIndex()) {
                if (i == j) {
                    continue
                }
                
                if (given > giveAndTake[j][i]) {
                    received[i]++
                } else if (given < giveAndTake[j][i]) {
                    continue
                } else if (giftDegree[i] > giftDegree[j]) {
                    received[i]++
                } else if (giftDegree[j] < giftDegree[i]) {
                    received[j]++
                }
            }
        }
        
        return received.maxOrNull()!!
    }
}